### PR TITLE
Support Kern and ForegroundColorFromContext as well as fix TruncatedL…

### LINKED ISSUE #1055: Support kCTKernAttributeName
Support this in the DWrite branch.

### PR DESCRIPTION
**Pull request by aballway**
…ine test

Fully support kCTKernAttributeName and kCTForegroundColorFromContextAttributeName.  Also fixes CTCatalog truncated line bug where the copied line would be misaligned.

Fixes #1055 and #1056

### CODE REVIEW

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm [hunk 1]
```diff
@@ -35,7 +35,7 @@ - (void)drawRect:(CGRect)rect {
 
     // Flips y-axis for our frame
     CGContextScaleCTM(context, 1.0f, -1.0f);
-    CGContextSetTextPosition(context, 0.0, 10.0);
+    CGContextSetTextPosition(context, 0.0, 0.0);
```
Comment (ms-jihua) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm]: ?

File: tests/unittests/CoreText/CTParagraphStyleTests.mm [hunk 2]
```diff
@@ -88,4 +88,35 @@
 
     success = CTParagraphStyleGetValueForSpecifier(paragraphStyleCopy, 513, sizeof(CGFloat), nullptr);
     ASSERT_TRUE_MSG(!success, "FAILED: An invalid specifier crash should not occur.");
+}
+
+TEST(CoreText, KernAttribute) {
+    const static float c_errorDelta = 0.001f;
+    NSMutableAttributedString* string = [[NSMutableAttributedString alloc] initWithString:@"TESTTESTTEST"];
+    [string autorelease];
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTParagraphStyleTests.mm]: nit: just peg this onto the line above - string = [[[alloc]init] autorelease];

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm (hunk 1 above)
Comment (aballway) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm]: Changed this when I was trying to find out where the problem was coming from.  Works either way, and iirc Nithish wanted me to change the hardcoded 10 (though 0 isn't much better)

**Comment by ms-jihua** [comment]
:shipit:

File: Frameworks/CoreText/CTRun.mm [hunk 3]
```diff
@@ -252,9 +252,18 @@ void _CTRunDraw(CTRunRef run, CGContextRef ctx, CFRange textRange, bool adjustTe
...
     id fontColor = [curRun->_attributes objectForKey:(id)kCTForegroundColorAttributeName];
     if (fontColor == nil) {
-        fontColor = [_LazyUIColor blackColor];
+        CFBooleanRef useContextColor =
+            static_cast<CFBooleanRef>([curRun->_attributes objectForKey:(id)kCTForegroundColorFromContextAttributeName]);
+        if (useContextColor && CFBooleanGetValue(useContextColor)) {
+            // Leave the context fill color as is
+        } else {
+            fontColor = [_LazyUIColor blackColor];
+            CGContextSetFillColorWithColor(ctx, reinterpret_cast<CGColorRef>(fontColor));
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTRun.mm]: Does this behavior match with reference platform?

Comment (aballway) [inline Frameworks/CoreText/CTRun.mm]: Yep!

**Comment by rajsesh-msft** [comment]
:shipit:

**Comment by MSFTFox** [comment]
>     // TODO::

should this TODO have an associated issue?

---
Refers to: Frameworks/CoreText/CTRun.mm:251 in e1afcff. [](commit_id = e1afcff6a3db7d36d69fe43d144deb8991c39652, deletion_comment = False)

File: Frameworks/CoreText/CTRun.mm [hunk 4]
```diff
@@ -252,9 +252,18 @@ void _CTRunDraw(CTRunRef run, CGContextRef ctx, CFRange textRange, bool adjustTe
     // Fix this once CTFont and UIFont become bridgable classes.
     id fontColor = [curRun->_attributes objectForKey:(id)kCTForegroundColorAttributeName];
     if (fontColor == nil) {
-        fontColor = [_LazyUIColor blackColor];
+        CFBooleanRef useContextColor =
+            static_cast<CFBooleanRef>([curRun->_attributes objectForKey:(id)kCTForegroundColorFromContextAttributeName]);
+        if (useContextColor && CFBooleanGetValue(useContextColor)) {
+            // Leave the context fill color as is
+        } else {
```
Comment (MSFTFox) [inline Frameworks/CoreText/CTRun.mm]: Why have an empty true case? Why not just !(useContextColor && CFBooleanGetValue(useContextColor))?

**Comment by aballway** [comment]
>     // TODO::

They should be part of future work, but I'm not certain.  This one should've been part of future work, but has been worked around since there's no reference to UIFont or CTFontRef here...

---
In reply to: [251223866](https://github.com/Microsoft/WinObjC/pull/1067#issuecomment-251223866) [](ancestors = 251223866)

---
Refers to: Frameworks/CoreText/CTRun.mm:251 in e1afcff. [](commit_id = e1afcff6a3db7d36d69fe43d144deb8991c39652, deletion_comment = False)

**Comment by MSFTFox** [comment]
:shipit:

**Pull request merged**